python/hail/methods/statgen.py

```diff
@@ -2186,28 +2184,94 @@ def balding_nichols_model(n_populations, n_samples, n_variants, n_partitions=Non
+    for name, var in {"populations": n_populations,
+                      "samples": n_samples,
+                      "variants": n_variants,
+                      "partitions": n_partitions}.items():
+        if var < 1:
+            raise ValueError("n_{} must be positive, got {}".format(name, var))
+
+    for name, var in {"pop_dist": pop_dist, "fst": fst}.items():
+        if len(var) != n_populations:
+            raise ValueError("{} must be of length n_populations={}, got length {}"
+                             .format(name, n_populations, len(var)))
+
+    if any(map(lambda x: x < 0, pop_dist)):
+        raise ValueError("pop_dist must be non-negative, got {}"
+                         .format(pop_dist))
+
+    if any(map(lambda x: x <= 0 or x >= 1, fst)):
+        raise ValueError("elements of fst must satisfy 0 < x < 1, got {}"
+                         .format(fst))
+
+    # verify af_dist and seed
+    if not af_dist._is_constant:
+        raise ExpressionException('balding_nichols_model expects af_dist to ' +
+                                  'have constant arguments: found expression ' +
+                                  'from source {}'
+                                  .format(af_dist._indices.source))
+
+    if not isinstance(af_dist._ir, hl.ir.ApplySeeded) or af_dist.dtype != tfloat64:
+        raise ValueError("af_dist must be a random function with return type tfloat64.")
+
+    fn = af_dist._ir.function
+    import inspect
+    params = eval('list(inspect.signature(hl.{}).parameters.keys())'.format(fn))
+    args = list(map(lambda arg: construct_expr(arg, tfloat64).value, af_dist._ir.args))
+    assert params[-1] == 'seed'
+    params = params[:-1]
+    struct_af_dist = hl.struct(type=fn, **{p: a for p, a in zip(params, args)}, seed=af_dist._ir.seed)
+
+    info("balding_nichols_model: generating genotypes for {} populations, {} samples, and {} variants..."
+         .format(n_populations, n_samples, n_variants))
+
+    # generate matrix table
+    rand = hl.utils.HailSeedGenerator(seed)
+
+    bn = hl.utils.range_matrix_table(n_variants, n_samples, n_partitions)
+    bn = bn.annotate_globals(n_populations=n_populations,
+                             n_samples=n_samples,
+                             n_variants=n_variants,
+                             n_partitions=n_partitions,
+                             pop_dist=pop_dist,
+                             fst=fst,
+                             ancestral_af_dist=struct_af_dist,
+                             seed=seed,
+                             mixture=mixture)
+    # col info
+    pop_f = hl.rand_dirichlet if mixture else hl.rand_cat
+    bn = bn.key_cols_by(sample_idx=bn.col_idx)
+    bn = bn.select_cols(pop=pop_f(pop_dist, seed=rand.next_seed()))
+
+    # row info
+    bn = bn.key_rows_by(locus=hl.locus_from_global_position(bn.row_idx, reference_genome=reference_genome),
+                        alleles=['A', 'C'])
+    bn = bn.select_rows(ancestral_af=af_dist,
```

Comment:
I'm a bit confused by the seed parameter. I guess that gives me full control of the seeding, but I would have thought I'd control seeding by the context in which I use the expression i.e.:
```python
mt = mt.annotate_globals(x = hl.rand_unif(0, 1)) # draw once for this MT
mt = mt.annotate_rows(y = hl.rand_unif(0, 1)) # draw once for each row
mt = mt.annotate_cols(z = hl.rand_unif(0, 1)) # draw once for each col
mt = mt.annotate_entries(x = hl.rand_unif(0, 1)) # draw once for each entry

foo(hl.rand_unif(0, 1)) # I cannot know how many draws without inspecting the function

x = hl.rand_unif(0, 1).value # ensures I get one draw
z = hl.rand_unif(0, 1)
zs = [z for _ in range(10)]
zs = [z.value for z in zs] # 10 draws
```

Reply:
In particular, how many times do I draw from the af_dist in this `select_rows`? For the model, we want a separate draw for each row, but it's not clear to me how to achieve that if I have to specify the seed *outside* of this function.